File: sdk/src/main/java/eu/darken/porter/sdk/PorterApiProvider.java
package eu.darken.porter.sdk;

import static eu.darken.porter.protocol.PorterProtocol.DELIVERY_METHOD_GET_BINDER;
import static eu.darken.porter.protocol.PorterProtocol.DELIVERY_METHOD_SEND_BINDER;

import android.content.BroadcastReceiver;
import android.content.ContentProvider;
import android.content.ContentValues;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.ProviderInfo;
import android.database.Cursor;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.IBinder;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * Receives the binder the Porter server sends when the app process starts. The SDK declares this
 * provider itself at {@code ${applicationId}.porter.api}; an app that must not expose it removes
 * the declaration with {@code tools:node="remove"}.
 *
 * <p>{@code android:permission} has to be one granted to the shell but not to normal apps, so that
 * only this app and the server can reach it; {@code android:exported} has to be true for the server
 * to reach it at all; {@code android:multiprocess} has to be false because the server reads the uid
 * once, when the app starts.
 *
 * <p>If the app runs in several processes, see {@link #enableMultiProcessSupport(boolean)}.
 */
public class PorterApiProvider extends ContentProvider {

    private static final String TAG = "PorterApiProvider";

    public static final String ACTION_BINDER_RECEIVED = "eu.darken.porter.sdk.action.BINDER_RECEIVED";

    private static boolean enableMultiProcess = false;

    private static boolean isProviderProcess = false;

    /**
     * Enables the built-in multi-process support. Call this as early as possible, for instance in a
     * static block of the Application class.
     */
    public static void enableMultiProcessSupport(boolean isProviderProcess) {
        Log.d(TAG, "Enable built-in multi-process support (from "
                + (isProviderProcess ? "provider process" : "non-provider process") + ")");

        PorterApiProvider.isProviderProcess = isProviderProcess;
        PorterApiProvider.enableMultiProcess = true;
    }

    /**
     * Asks for the binder in a process that does not host the provider;
     * {@link #enableMultiProcessSupport(boolean)} must have been called first.
     */
    public static void requestBinderForNonProviderProcess(@NonNull Context context) {
        if (isProviderProcess) {
            return;
        }

        Log.d(TAG, "request binder in non-provider process");

        // Below API 33 a registered receiver is exported, so any app can send this action to us.
        // Treat the broadcast as a notification only and read the binder from the provider, which
        // android:permission and the same-uid exemption restrict to this app and the server.
        BroadcastReceiver receiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                Log.i(TAG, "binder announced by broadcast");
                fetchBinderFromProvider(context);
            }
        };

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            context.registerReceiver(receiver, new IntentFilter(ACTION_BINDER_RECEIVED), Context.RECEIVER_NOT_EXPORTED);
        } else {
            context.registerReceiver(receiver, new IntentFilter(ACTION_BINDER_RECEIVED));
        }

        // A delivery this process refused was refused because a connection was live. Once that
        // connection dies nothing is live, the refusal no longer applies, and the binder the
        // provider process holds can be adopted. It runs as a post-death hook rather than as a dead
        // listener so that every listener registered for that death is told before the replacement
        // announces itself: an app that registers its own listeners after this call would otherwise
        // sit behind this one.
        Context appContext = context.getApplicationContext();
        Porter.addPostBinderDeadHook(() -> fetchBinderFromProvider(appContext));

        fetchBinderFromProvider(context);
    }

    /**
     * Asks every authority a server can have delivered to, because the session this process is
     * looking for is on whichever backend the provider process attached.
     */
    static boolean fetchBinderFromProvider(@NonNull Context context) {
        if (fetchThrough(context, PorterProtocolDelivery.INSTANCE)) return true;
        // ShizukuProtocolDelivery names the container as a type, so the class cannot load at all
        // where the optional artifact is absent.
        return ShizukuCompat.isPresent() && fetchThrough(context, ShizukuProtocolDelivery.INSTANCE);
    }

    static boolean fetchThrough(@NonNull Context context, @NonNull PorterDelivery delivery) {
        Bundle reply;
        try {
            reply = context.getContentResolver().call(
                    Uri.parse("content://" + context.getPackageName() + delivery.authoritySuffix()),
                    DELIVERY_METHOD_GET_BINDER, null, new Bundle());
        } catch (Throwable tr) {
            reply = null;
        }

        if (reply == null) return false;
        IBinder binder = delivery.readBinder(reply);
        if (binder == null) return false;

        Log.i(TAG, "Binder received from other process");
        // Ungated on purpose: this is the provider process's own connection, on the backend that
        // process already selected. Rejecting it here would refuse a binder the app is using.
        PorterSession.adoptBackend(delivery.backend());
        PorterSession.onBinderReceived(binder, context.getPackageName(), delivery.backend());
        return true;
    }

    /** The envelope this provider speaks; a subclass overrides it to answer another authority. */
    @NonNull
    PorterDelivery delivery() {
        return PorterProtocolDelivery.INSTANCE;
    }

    @Override
    public void attachInfo(Context context, ProviderInfo info) {
        super.attachInfo(context, info);

        if (info.multiprocess)
            throw new IllegalStateException("android:multiprocess must be false");

        if (!info.exported)
            throw new IllegalStateException("android:exported must be true");

        isProviderProcess = true;
    }

    @Override
    public boolean onCreate() {
        return true;
    }

    @Nullable
    @Override
    public Bundle call(@NonNull String method, @Nullable String arg, @Nullable Bundle extras) {
        if (extras == null) {
            return null;
        }

        Bundle reply = new Bundle();
        switch (method) {
            case DELIVERY_METHOD_SEND_BINDER: {
                handleSendBinder(extras);
                break;
            }
            case DELIVERY_METHOD_GET_BINDER: {
                if (!handleGetBinder(reply)) {
                    return null;
                }
                break;
            }
        }
        return reply;
    }

    private void handleSendBinder(@NonNull Bundle extras) {
        if (Porter.pingBinder()) {
            Log.d(TAG, "sendBinder is called when already a living binder");
            return;
        }

        IBinder binder = delivery().readBinder(extras);
        if (binder == null) {
            Log.w(TAG, "sendBinder is called without a binder");
            return;
        }

        Log.d(TAG, "binder received");

        Porter.onBinderReceived(
                getContext(), binder, getContext().getPackageName(), delivery().backend());

        if (enableMultiProcess) {
            Log.d(TAG, "broadcast binder");

            Intent intent = new Intent(ACTION_BINDER_RECEIVED)
                    .setPackage(getContext().getPackageName());
            getContext().sendBroadcast(intent);
        }
    }

    private boolean handleGetBinder(@NonNull Bundle reply) {
        // Other processes in the same app can read the provider without permission
        IBinder binder = PorterSession.binderFor(delivery().backend());
        if (binder == null) {
            return false;
        }

        delivery().writeBinder(reply, binder);
        return true;
    }

    // no other provider methods
    @Nullable
    @Override
    public final Cursor query(@NonNull Uri uri, @Nullable String[] projection, @Nullable String selection,
                              @Nullable String[] selectionArgs, @Nullable String sortOrder) {
        return null;
    }

    @Nullable
    @Override
    public final String getType(@NonNull Uri uri) {
        return null;
    }

    @Nullable
    @Override
    public final Uri insert(@NonNull Uri uri, @Nullable ContentValues values) {
        return null;
    }

    @Override
    public final int delete(@NonNull Uri uri, @Nullable String selection, @Nullable String[] selectionArgs) {
        return 0;
    }

    @Override
    public final int update(@NonNull Uri uri, @Nullable ContentValues values, @Nullable String selection,
                            @Nullable String[] selectionArgs) {
        return 0;
    }
}

File: sdk/src/main/java/eu/darken/porter/sdk/PorterApiProvider.kt
package eu.darken.porter.sdk

import android.app.Application
import android.content.BroadcastReceiver
import android.content.ComponentName
import android.content.ContentProvider
import android.content.ContentValues
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.content.pm.ProviderInfo
import android.database.Cursor
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.util.Log
import androidx.annotation.VisibleForTesting
import eu.darken.porter.protocol.PorterProtocol.DELIVERY_METHOD_GET_BINDER
import java.io.File
import java.io.IOException
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Receives the binder the Porter server sends when the app process starts. The SDK declares this
 * provider itself at `${applicationId}.porter.api`; an app that must not expose it removes the
 * declaration with `tools:node="remove"`.
 *
 * `android:permission` has to be one granted to the shell but not to normal apps, so that only
 * this app and the server can reach it; `android:exported` has to be true for the server to reach
 * it at all; `android:multiprocess` has to be false because the server reads the uid once, when
 * the app starts.
 *
 * If the app runs in several processes, see [requestBinderForNonProviderProcess].
 *
 * A subclass can answer further methods in [call] and hand the rest to `super.call`; nothing else
 * can be overridden.
 */
public open class PorterApiProvider : ContentProvider() {

    private val endpoint = DeliveryEndpoint(PorterProtocolDelivery)

    final override fun attachInfo(context: Context, info: ProviderInfo) {
        super.attachInfo(context, info)
        endpoint.attached(info)
    }

    final override fun onCreate(): Boolean = true

    override fun call(method: String, arg: String?, extras: Bundle?): Bundle? =
        endpoint.call(requireContext(), method, extras)

    // no other provider methods
    final override fun query(
        uri: Uri,
        projection: Array<String>?,
        selection: String?,
        selectionArgs: Array<String>?,
        sortOrder: String?,
    ): Cursor? = null

    final override fun getType(uri: Uri): String? = null

    final override fun insert(uri: Uri, values: ContentValues?): Uri? = null

    final override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?): Int = 0

    final override fun update(uri: Uri, values: ContentValues?, selection: String?, selectionArgs: Array<String>?): Int = 0

    public companion object {

        private const val TAG = "PorterApiProvider"

        internal const val ACTION_BINDER_RECEIVED: String = "eu.darken.porter.sdk.action.BINDER_RECEIVED"

        /** Set by a provider of this SDK attaching in this process. */
        @Volatile
        internal var isProviderProcess = false

        /** Guards [registered]. */
        private val registration = Any()

        /** Whether this process already listens for announcements and deaths. */
        private var registered = false

        /** Set while a fetch is queued and has not started, so a burst of requests queues one. */
        private val fetchQueued = AtomicBoolean(false)

        /**
         * Obtains the connection in a process that does not host the provider, and keeps obtaining
         * the next one whenever the provider process accepts it. The result arrives on
         * [Porter.connection]; nothing here waits for it, so any thread may call this, and calling
         * it again only asks the provider process once more.
         *
         * In the process that hosts the provider this does nothing: the server delivers there.
         */
        public fun requestBinderForNonProviderProcess(context: Context) {
            val appContext = context.applicationContext ?: context
            if (hostsProvider(appContext)) return

            synchronized(registration) {
                if (!registered) {
                    register(appContext)
                    registered = true
                }
            }

            scheduleFetch(appContext)
        }

        private fun register(appContext: Context) {
            Log.d(TAG, "listening for the provider process's binder")

            // Below API 33 a registered receiver is exported, so any app can send this action to us.
            // Treat the broadcast as a notification only and read the binder from the provider, which
            // android:permission and the same-uid exemption restrict to this app and the server.
            val receiver = object : BroadcastReceiver() {
                override fun onReceive(context: Context, intent: Intent) {
                    Log.i(TAG, "binder announced by broadcast")
                    scheduleFetch(appContext)
                }
            }

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                appContext.registerReceiver(receiver, IntentFilter(ACTION_BINDER_RECEIVED), Context.RECEIVER_NOT_EXPORTED)
            } else {
                appContext.registerReceiver(receiver, IntentFilter(ACTION_BINDER_RECEIVED))
            }

            // A delivery this process refused was refused because a connection was live. Once that
            // connection dies nothing is live, the refusal no longer applies, and the binder the
            // provider process holds can be adopted. It runs as a post-death hook so that every
            // collector sees that death before the replacement announces itself.
            Porter.addPostBinderDeadHook { scheduleFetch(appContext) }
        }

        /** Fetches off the calling thread: the attach behind it blocks, for seconds on the Shizuku wire. */
        private fun scheduleFetch(appContext: Context) {
            if (!fetchQueued.compareAndSet(false, true)) return
            Porter.deliveryExecutor.execute {
                // Cleared before the fetch, so an announcement that arrives during it queues another.
                fetchQueued.set(false)
                fetchBinderFromProvider(appContext)
            }
        }

        /**
         * Whether this process is the one the provider runs in. A provider that attached here says so
         * directly; before it attaches, the declared process name is compared with this process's.
         */
        private fun hostsProvider(context: Context): Boolean {
            if (isProviderProcess) return true
            val current = currentProcessName() ?: return false
            val providers = listOf(PorterApiProvider::class.java, PorterShizukuApiProvider::class.java)
            return providers.any { provider ->
                val info = try {
                    context.packageManager.getProviderInfo(ComponentName(context, provider), 0)
                } catch (e: PackageManager.NameNotFoundException) {
                    null
                }
                // A provider without a process of its own runs in the application's.
                info != null && current == (info.processName ?: context.applicationInfo.processName)
            }
        }

        private fun currentProcessName(): String? =
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                Application.getProcessName()
            } else {
                try {
                    File("/proc/self/cmdline").readText().substringBefore('\u0000').takeIf { it.isNotEmpty() }
                } catch (e: IOException) {
                    null
                }
            }

        /**
         * Asks every authority a server can have delivered to, because the connection this process
         * is looking for is on whichever backend the provider process attached. Blocks for the attach.
         */
        internal fun fetchBinderFromProvider(context: Context): Boolean {
            if (fetchThrough(context, PorterProtocolDelivery)) return true
            // ShizukuProtocolDelivery names the container as a type, so the class cannot load at all
            // where the optional artifact is absent.
            return ShizukuCompat.isPresent() && fetchThrough(context, ShizukuProtocolDelivery)
        }

        internal fun fetchThrough(context: Context, delivery: PorterDelivery): Boolean {
            val reply = try {
                context.contentResolver.call(
                    Uri.parse("content://" + context.packageName + delivery.authoritySuffix),
                    DELIVERY_METHOD_GET_BINDER, null, Bundle(),
                )
            } catch (tr: Throwable) {
                null
            } ?: return false

            val binder = delivery.readBinder(reply) ?: return false

            Log.i(TAG, "Binder received from other process")
            // Ungated on purpose: this is the provider process's own connection, on the backend that
            // process already selected. Rejecting it here would refuse a binder the app is using.
            Porter.adoptBackend(delivery.backend)
            Porter.onBinderReceived(binder, context.packageName, delivery.backend)
            return true
        }

        /** Forgets the registration and the provider flag, so one test cannot see another's. */
        @VisibleForTesting
        internal fun resetForTest() {
            synchronized(registration) {
                registered = false
            }
            fetchQueued.set(false)
            isProviderProcess = false
        }
    }
}
